content/en/blog/fsm2/index.md

```diff
@@ -1,207 +1,357 @@
+interesting commercial high-end tools supported correctness verification using
+proofing theorems.
+
+I've continued studying the SCXML, and it might still be an option in the
+future. It's standard, and there are good tools for it. Unfortunately, proper
+open-source tools are still missing. So, we decided to build a simple
+language with tools in Unix and build something as fast as possible. The
+result is better than we could have guessed -- and, of course, it proved our
+hypothesis.
+
+### Lua option
+
+During these searches, I discovered [a Go-native implementation of
+Lua](https://github.com/Shopify/go-lua), which was open-source. Luckily, Lua is
+now included as an embedded scripting language *for our FSM*. It can be used to
+[implement custom
+triggers](https://findy-network.github.io/blog/2023/03/13/no-code-ssi-chatbots-part-i/#event)
+and custom output events.
+
+You can write Lua directly to YAML files or include a file link and write Lua
+scripts to external files. (*Depending on the deployment model, that could be
+a security problem, but we are in an early stage of the technology, and it will be
+solved when we have a correctness tool integrated.*)
+
+I saw during the studies that all-in for the *embedded* Lua would have brought
+at least the following challenges:
+1. Whole new API and bridge to Aries protocols would be needed. A simple C/S API
+   wouldn't be enough, but a notification system would also be required.
+1. Lua is a Turing-complete language (correctness guarantees, the halting problem,
+   etc.)
+1. What extra would we be brought in when compared just to offer Lua stubs to our
+   current gRPC API?
+
+## The Event Processor 
+
+Our FSM engine follows the exact same principle as [the SCXML processor](https://en.wikipedia.org/wiki/SCXML) does:
+
+> An SCXML processor is a pure event processor. The only way to get data into an
+> SCXML state machine is to send external events to it. The only way to get data
+> out is to receive events from it.
+
+In the software industry, there are some other similar systems, like, for example,
+[eBPF](https://ebpf.io/), where correctness is provided without formal language.
+eBPF automatically **rejects programs without strong exit guarantees**, i.e.,
+for/while loops without exit conditions. That's achieved with static code
+analysis, which allows using conventional programming languages.
+
+Now that *we have brought Lua-scripting into our FSMs*, we should also get
+*code analysis for the exit guarantees*, but let's first figure out how
+good a fit it is overall. My first impressions have been very good.
+
+
+## Services and Conversations
+
+The drawing below includes our multi-tenant model. FSM instances and their
+memory containers are on the right-hand side of the diagram. The smaller ones
+are conversation instances that hold the state of each pairwise connection
+status. The larger one is *the service state machine*, which isn’t needed on all
```

Comment:
Modify:

... The larger one is *the service state machine*. <chapter_ends>

<chapter_starts>: The service state machine isn't needed on all chatbot applications. It is needed if the SSI agent...